File: Max_List.js
#!/usr/bin/env node
// Max_List.js - Node.js port of Max_List.java
// Usage: interactive tokens. Enter numbers one per token, 'M' to show full stats and exit, 'N' to show minimum and exit.
const readline = require('readline');

function fmt(d) {
  if (!Number.isFinite(d)) return String(d);
  let s = d.toFixed(6);
  // trim trailing zeros and optional trailing dot
  s = s.replace(/(\.\d*?)0+$/, '$1');
  s = s.replace(/\.$/, '');
  return s;
}

function computeModes(freq) {
  let best = 0;
  const modes = [];
  for (const k of Object.keys(freq)) {
    const v = freq[k];
    if (v > best) { best = v; modes.length = 0; modes.push(Number(k)); }
    else if (v === best) modes.push(Number(k));
  }
  return { best, modes };
}

function processTokens(tokens, state, jsonMode) {
  for (const tokenRaw of tokens) {
    const token = tokenRaw.trim();
    if (token === '') continue;

    if (token.toUpperCase() === 'M') {
      if (!state.haveNumber) {
        if (jsonMode) console.log(JSON.stringify({ error: 'No numbers were entered.' }));
        else console.log('No numbers were entered.');
        return { done: true };
      }
      const average = state.mean;
      const variance = (state.count > 0) ? (state.m2 / state.count) : 0.0;
      const stddev = Math.sqrt(variance);

      if (jsonMode) {
        const { best, modes } = computeModes(state.freq);
        const out = { maximum: state.max, minimum: state.min, sum: state.sum, count: state.count, average, variance, stddev, modes: (best <= 1) ? [] : modes, modeFrequency: (best <= 1) ? 0 : best };
        console.log(JSON.stringify(out));
      } else {
        console.log('Maximum = ' + fmt(state.max));
        console.log('Minimum = ' + fmt(state.min));
        console.log('Sum = ' + fmt(state.sum));
        console.log('Count = ' + state.count);
        console.log('Average = ' + fmt(average));
        console.log('Variance (population) = ' + fmt(variance));
        console.log('Standard deviation (population) = ' + fmt(stddev));

        const { best, modes } = computeModes(state.freq);
        if (best <= 1) console.log('Mode: none (all values are unique)');
        else {
          process.stdout.write('Mode(s) (frequency=' + best + '): ');
          for (let i = 0; i < modes.length; i++) {
            process.stdout.write(fmt(modes[i]));
            if (i < modes.length - 1) process.stdout.write(', ');
          }
          process.stdout.write('\n');
        }
      }
      return { done: true };
    }

    if (token.toUpperCase() === 'N') {
      if (jsonMode) {
        if (!state.haveNumber) console.log(JSON.stringify({ error: 'No numbers were entered.' }));
        else console.log(JSON.stringify({ minimum: state.min }));
      } else {
        if (!state.haveNumber) console.log('No numbers were entered.');
        else console.log('Minimum = ' + fmt(state.min));
      }
      return { done: true };
    }

    const val = Number(token);
    if (Number.isNaN(val)) {
      if (!jsonMode) console.log("Invalid input. Enter a number, or 'M'/'N' to finish.");
      continue;
    }

    if (!state.haveNumber) {
      state.max = val; state.min = val; state.haveNumber = true;
    } else {
      if (val > state.max) state.max = val;
      if (val < state.min) state.min = val;
    }

    state.count += 1;
    state.sum += val;
    const delta = val - state.mean;
    state.mean += delta / state.count;
    state.m2 += delta * (val - state.mean);

    const key = String(val);
    state.freq[key] = (state.freq[key] || 0) + 1;
  }
  return { done: false };
}

// CLI parsing
const argv = process.argv.slice(2);
let jsonMode = false;
const args = [];
for (const a of argv) {
  if (a === '--json' || a === '-j') jsonMode = true;
  else args.push(a);
}

const state = { max: -Infinity, min: Infinity, haveNumber: false, count: 0, sum: 0.0, mean: 0.0, m2: 0.0, freq: {} };

if (args.length > 0) {
  const res = processTokens(args, state, jsonMode);
  if (res.done) process.exit(0);
}

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
rl.setPrompt('> ');
rl.prompt();
rl.on('line', (line) => {
  const tokens = line.trim().split(/\s+/);
  const res = processTokens(tokens, state, jsonMode);
  if (res.done) { rl.close(); return; }
  rl.prompt();
}).on('close', () => {
  if (!state.haveNumber) {
    if (jsonMode) console.log(JSON.stringify({ info: 'No more input. Exiting.' }));
    else console.log('No more input. Exiting.');
  }
  process.exit(0);
});

File: Max_List.java
import java.util.Scanner;
import java.util.Map;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;

public class Max_List {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter numbers one by one. Enter 'M' to output statistics (max,min,sum,count,average,variance,stddev,mode), or 'N' to output the minimum and exit.");

        double max = Double.NEGATIVE_INFINITY;
        double min = Double.POSITIVE_INFINITY;
        boolean haveNumber = false;

        long count = 0;
        double sum = 0.0;
        double mean = 0.0; // for Welford
        double m2 = 0.0;   // for Welford
        Map<Double, Integer> freq = new HashMap<>();

        while (true) {
            System.out.print("> ");
            if (!sc.hasNext()) { // EOF
                System.out.println("No more input. Exiting.");
                break;
            }
            String token = sc.next();

            if (token.equalsIgnoreCase("M")) {
                if (haveNumber) {
                    double average = mean;
                    double variance = (count > 0) ? (m2 / count) : 0.0; // population variance
                    double stddev = Math.sqrt(variance);

                    System.out.println("Maximum = " + fmt(max));
                    System.out.println("Minimum = " + fmt(min));
                    System.out.println("Sum = " + fmt(sum));
                    System.out.println("Count = " + count);
                    System.out.println("Average = " + fmt(average));
                    System.out.println("Variance (population) = " + fmt(variance));
                    System.out.println("Standard deviation (population) = " + fmt(stddev));

                    // mode(s)
                    int best = 0;
                    List<Double> modes = new ArrayList<>();
                    for (Map.Entry<Double, Integer> e : freq.entrySet()) {
                        int f = e.getValue();
                        if (f > best) {
                            best = f;
                            modes.clear();
                            modes.add(e.getKey());
                        } else if (f == best) {
                            modes.add(e.getKey());
                        }
                    }
                    if (best <= 1) {
                        System.out.println("Mode: none (all values are unique)");
                    } else {
                        System.out.print("Mode(s) (frequency=" + best + "): ");
                        for (int i = 0; i < modes.size(); i++) {
                            System.out.print(fmt(modes.get(i)));
                            if (i < modes.size() - 1) System.out.print(", ");
                        }
                        System.out.println();
                    }
                } else {
                    System.out.println("No numbers were entered.");
                }
                break;
            }

            if (token.equalsIgnoreCase("N")) {
                if (haveNumber) {
                    System.out.println("Minimum = " + fmt(min));
                } else {
                    System.out.println("No numbers were entered.");
                }
                break;
            }

            try {
                double val = Double.parseDouble(token);
                if (!haveNumber) {
                    max = min = val;
                    haveNumber = true;
                } else {
                    if (val > max) max = val;
                    if (val < min) min = val;
                }

                // update counters and Welford for variance
                count++;
                sum += val;
                double delta = val - mean;
                mean += delta / count;
                m2 += delta * (val - mean);

                // frequency for mode
                freq.put(val, freq.getOrDefault(val, 0) + 1);

            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Enter a number, or 'M'/'N' to finish.");
            }
        }

        sc.close();
    }

    private static String fmt(double d) {
        if (Double.isInfinite(d) || Double.isNaN(d)) return Double.toString(d);
        // show up to 6 decimal places but trim trailing zeros
        String s = String.format("%.6f", d);
        if (s.indexOf('.') >= 0) {
            while (s.endsWith("0")) {
                s = s.substring(0, s.length() - 1);
            }
            if (s.endsWith(".")) s = s.substring(0, s.length() - 1);
        }
        return s;
    }
}
